File: 20200223/python/baseballgame_test.py
from itertools import permutations


def solution(baseball):
    numbers = [str(x) for x in range(1, 10)]
    candidates = [''.join(result_set) for result_set in permutations(numbers, 3)]

    for (question, strikes, balls) in baseball:
        candidates = [candidate for candidate in candidates if match(str(question), candidate) == (strikes, balls)]

    return len(candidates)


def match(question, candidate):
    strikes, balls = 0, 0

    for i in range(3):
        if question[i] == candidate[i]:
            strikes += 1
        elif question[i] in candidate:
            balls += 1

    return strikes, balls


def test_solution():
    assert solution([[123, 1, 1], [356, 1, 0], [327, 2, 0], [489, 0, 1]]) == 2


def test_match():
    assert match('123', '324') == (1, 1)

File: 20200223/js/baseballgame.test.js
const solution = (baseball) =>
  Array(900).fill(100).map((v, i) => `${v + i}`)
    .filter(v => !v.includes('0') && [...new Set(v)].length === 3)
    .filter(v => isCandidate(v, baseball)).length;

const isCandidate = (candidate, baseball) => baseball.every(game => check(candidate, game));

const check = (candidate, [question, strikes, balls]) => {
  const { strike, ball } = [...`${question}`].reduce(({ strike, ball }, c, index) => {
    if (!candidate.includes(c)) {
      return { strike, ball };
    }
    return candidate.indexOf(c) === index
      ? { strike: strike + 1, ball: ball }
      : { strike: strike, ball: ball + 1 };
  }, { strike: 0, ball: 0 });

  return strikes === strike && balls === ball;
};

test('solution', () => {
  expect(solution([[123, 1, 1], [356, 1, 0], [327, 2, 0], [489, 0, 1]])).toBe(2);
});

test('isCandidate', () => {
  expect(isCandidate('324', [[123, 1, 1], [356, 1, 0], [327, 2, 0], [489, 0, 1]])).toBe(true);
});

test('check', () => {
  expect(check('324', [123, 1, 1])).toBe(true);
});
